migrations: VersionedRuntimeUpgrade

liamaharon opened the pull request:
Closes paritytech/substrate#13107

## Description

Make it easier to write versioned runtime upgrades.

VersionedRuntimeUpgrade allows developers to write migrations without worrying about checking and setting storage versions. Instead, the developer wraps their migration in this struct which takes care of version handling using best practices.

It takes 5 type parameters:
- `From`: The version being upgraded from
- `To`: The version being upgraded to
- `Inner`: An implementation of `OnRuntimeUpgrade`
- `Pallet`: The Pallet being upgraded
- `Weight`: The runtime's RuntimeDbWeight implementation

When a VersionedRuntimeUpgrades `on_runtime_upgrade` method is called, the on-chain version of the pallet is compared to `From`. If they match, `Inner::on_runtime_upgrade` is called and the pallets on-chain version is set to `To`. Otherwise, a warning is logged notifying the developer that the upgrade was a noop and should probably be removed.

## Example

Example using this for a migration in the Polkadot repo: https://github.com/paritytech/polkadot/compare/liam-versioned-runtime-upgrade-example

## Decision not to remove `#[pallet::storage_version]`

The possibility of removing `#[pallet::storage_version]` leaving only an on-chain version has been discussed. Although this would be nice, I think it is necessary to keep so that we may correctly set the initial on-chain version of a pallet when it is first initialized. 

See https://github.com/paritytech/polkadot-sdk/issues/109 for more which I'm planning to work on soon, which appears to require `#[pallet::storage_version]`.

## Decision to use `From` and `To` rather than just `Version` (`From`)

See https://github.com/paritytech/substrate/pull/14311#discussion_r1221399650

## TODOs
- [x] Add some tests making sure the Inner hooks are called when the version matches, and not called when the version doesn't match
- [x] Test that bytes are passed correctly from pre_upgrade to post_upgrade
- [ ] Update CI to run `experimental` tests https://github.com/paritytech/substrate/issues/14345#issuecomment-1604097398

liamaharon committed:
VersionedRuntimeUpgrade

liamaharon committed:
only require one version and add a pre-upgrade check

liamaharon committed:
add docs

liamaharon committed:
improve warning log

liamaharon committed:
improve comments

liamaharon committed:
fix log

ggwpez commented on frame/support/src/migrations.rs:
```diff
@@ -15,19 +15,126 @@
+/// takes care of version handling using best practices.
+///
+/// It takes 4 type parameters:
+/// - `Version`: The version being upgraded *from*.
```
This is a good start, but I think we need `from` and `to`.  
It should advance exactly from `from` to `to`, so we can lock it down. No more `<` or `>` squishy logic anymore.

liamaharon replied:
I originally implemented this with `From` and `To`, but refactored to only `Version` after thinking that `To` is redundant.

I see `To` as only useful to support these two cases, which I can't think of scenarios where they would be best practice:
1. Skipping version numbers (even the nom pools implemented `MigrateV3ToV5` using two migrations, each incrementing one version: https://github.com/paritytech/substrate/blob/aeb8c3107f79afef081dde2b9aa2866a6596f6ba/frame/nomination-pools/src/migration.rs#L497 )
2. A single migration running on more than one version - this seems strange, and probably means the migration should be rearchitected

Let me know if you can think of any scenarios I've missed, or if you think we should support any of the above cases, and I'm happy to switch back to `To`/`From`.

thiolliere replied:
> 2\. A single migration running for subsequent upgrades - this seems strange and unnecessary. I can't think of cases where it would be needed.

what if somebody wants to implement a migration from version 2 to version 4 in a more efficient way than executing version 2 to 3 and version 3 to 4 ?
Also we could provide different type. One type for upgrade from one version to next version, one type for upgrade from one version to another version.

ggwpez replied:
Yes we actually have this in the scheduler: `1->4`, `2->4` and nom pools also had a batched migration from `3->5`.  
I think making this explicit also helps with chaining them together later on.

>One type for upgrade from one version to next version, one type for upgrade from one version to another version.

Yea the former could be a wrapper around the latter.

ggwpez commented on frame/support/src/migrations.rs:
```diff
@@ -15,19 +15,126 @@
+/// It takes 4 type parameters:
+/// - `Version`: The version being upgraded *from*.
+/// - `Inner`: An implementation of `OnRuntimeUpgrade`.
+/// - `Pallet`: The Pallet being upgraded.
```
The main usage for this will be pallets, so it is probably find to design it pallet-centric.  
I am just thinking if we also want to use this for non-pallet migrations :thinking:

liamaharon replied:
Currently, versions are always relative to the on-chain pallet version. Are you thinking we could create on-chain versions for other arbitrary storage groups that we manage with runtime upgrades? Do you have an example?

ggwpez replied:
I was just thinking in the general spirit of "storage migrations". These are not necessarily related to pallets - but for our use cases they almost always are.  
Not sure if we can create a `GetVersion` trait or something that can be blanket implemented for the `PalletInfoAccess` trait or something to allow for non-pallet migrations, but it is probably fine for now. As we want to get this shipped :laughing:

ggwpez commented on frame/support/src/migrations.rs:
```diff
@@ -15,19 +15,126 @@
+/// the pallets storage matches `Version`. If the versions do not match, it writes a log notifying
+/// the developer that the migration is a noop.
+impl<
+		Version: Get<StorageVersion>,
```
Did you try out how it looks like with an associated constant?  
I think the `Get` always makes for an awkward usage :seedling: 

How would this be used btw?  
type MigrateToV3 = VersionedRuntimeUpgrade<ConstU32<3>, InnerMigrationToV3, Pallet, Weights>`?  
It is slightly odd that it is implemented for all `Version`, but not a problem.

ggwpez reviewed (commented):
Thanks for starting this!

liamaharon committed:
use associated constants

thiolliere commented:
Looks good,

I saw the example in https://github.com/paritytech/polkadot/compare/liam-versioned-runtime-upgrade-example
I wonder about usage: instead of using the type in the kusama runtime it should be used to declare the migration with version checked inside the pallet directly, no?
the pallet configuration would expose both `VersionedCheckedMigrateToV6` and `VersionedUncheckedMigrateToV6`:
```rust
type VersionCheckedMigrateToV6 = frame_support::migrations::VersionedRuntimeUpgrade<
			V5ToV6,
			parachains_configuration::migration::v6::VersionedUncheckedMigrateToV6<Runtime>,
			Configuration,
			RocksDbWeight,
		>,
```

liamaharon committed:
allow passing from and to versions

liamaharon committed:
test versioned_runtime_upgrade

liamaharon commented:
Thanks @ggwpez and @thiolliere for your comments. I've addressed them:
- Replaced `Version` with `To` and `From` for clearer behavior, and allowing skipping versions
- Updated my example on https://github.com/paritytech/polkadot/compare/liam-versioned-runtime-upgrade-example to define `UnversionedMigrateV5ToV6` and `VersionedMigrateV5ToV6` in the migrations file, and only use `VersionedMigrateV5ToV6` in the runtimes

and also written some tests.

Ready to review again when you have some time.

liamaharon committed:
fix typo

liamaharon committed:
improve docs

liamaharon commented:
Question: after merging, should we apply this to applicable `Unreleased` migrations?

liamaharon committed:
docs

liamaharon committed:
docs

kianenigma commented:
The main reason I suggested removing `#[pallet::storage_version]` is that I envision a single system-like migration pallet (possibly https://github.com/paritytech/substrate/pull/14275) to handle all the things that @thiolliere and @ggwpez discussed, like jumping migrations and so on.

From a DevEx perspective, I still want to suggest finding a way to not do this, as having both the onchain version and the fixed pallet version is not optimal.

kianenigma commented on frame/support/src/migrations.rs:
````diff
@@ -15,19 +15,128 @@
+/// that the upgrade was a noop and should probably be removed.
+///
+/// Example:
+/// ```ignore
````
possibly not to difficult to make this not be `ignore` and actually compile.

liamaharon replied:
are you thinking I'd use the dummy pallet and upgrades from my test file? I'd rather not use a migration from a real pallet since it creates a dependency that could break. I'm also curious what the benefit/s are of making the comment compile

kianenigma commented on frame/support/tests/versioned_runtime_upgrade.rs:
```diff
@@ -0,0 +1,231 @@
+type UncheckedExtrinsic = frame_system::mocking::MockUncheckedExtrinsic<Test>;
+type Block = frame_system::mocking::MockBlock<Test>;
+
+#[frame_support::pallet]
```
a dummy pallet only needs `Config` and `Pallet`.

kianenigma commented on frame/support/tests/versioned_runtime_upgrade.rs:
```diff
@@ -0,0 +1,231 @@
+
+type AccountId = u64;
+
+impl frame_system::Config for Test {
```
`derive_impl`?

kianenigma commented on frame/support/src/migrations.rs:
```diff
@@ -15,19 +15,128 @@
+		From: Get<u16>,
+		To: Get<u16>,
+		Inner: OnRuntimeUpgrade,
+		Pallet: GetStorageVersion<CurrentStorageVersion = StorageVersion> + PalletInfoAccess,
+		DbWeight: Get<RuntimeDbWeight>,
```
For readability purposes, I would consider moving `Pallet` and `DbWeight` to be associated types rather than generics and hardcoded. This might require you to add another trait. When a pallet is implementing this new trait, it will fix its `Pallet` and `DbWeight` (both of which can be obtained from `system`).

Then, in the runtime, when you are wrapping things in `VersionedRuntimeUpgrade`, you only see `(VersionedRuntimeUpgrade<1, 2, Foo>, VersionedRuntimeUpgrade<5, 6, Bar>)`. 

Possibly not worth it, especially as I am still critical of the fundamental approach.

liamaharon replied:
I'm able to remove the `DbWeight` from the runtime file like this, without needing to create a new trait:

before (inside migrations file):
```rust
pub type VersionCheckedMigrateV5ToV6<Runtime, Pallet, DbWeight> = VersionedRuntimeUpgrade<
	ConstU16<5>,
	ConstU16<6>,
	VersionUncheckedMigrateV5ToV6<Runtime>,
	Pallet,
	DbWeight,
>;
```

after (inside migrations file):
```rust
pub type VersionCheckedMigrateV5ToV6<Runtime, Pallet> = VersionedRuntimeUpgrade<
	ConstU16<5>,
	ConstU16<6>,
	VersionUncheckedMigrateV5ToV6<Runtime>,
	Pallet,
	<Runtime as frame_system::Config>::DbWeight,
>;
```

Then in the runtime file, you no longer need to pass the DbWeight:
```rust
parachains_configuration::migration::v6::VersionCheckedMigrateV5ToV6<
	Runtime,
	Configuration,
>,
```

Let me know if this approach is also OK, it seems to keep things clean in the runtime file without needing to create an extra trait, but I'm a noob with associated types so definitely may be missing something, please let me know if I am.

liamaharon commented:
Thanks for your comments @kianenigma. 

>> I think it is necessary to keep so that we may correctly set the initial on-chain version of a pallet when it is first initialized.

> can be solved in various other ways.
 
If we don't include the storage version in the pallet, then the only place the current pallet version would exist for a new chain is in the migration itself. 

I think that means that to ensure the on-chain version is set correctly we would need to 
- tightly couple at least the latest migration with the pallet 
- on initialization, use the `To` version on the latest migration to init the on-chain version
- potentially, this could handled by Oliver's new Migrations pallet
  - since migrations are tightly coupled with the pallet, the Migrations pallet could automatically check and execute these upgrades rather than us needing to list them in the `Executive` `Migrations` tuple

Is that close to what you're thinking or did you have something else in mind?

liamaharon committed:
remove event from dummy pallet

ggwpez commented:
>From a DevEx perspective, I still want to suggest finding a way to not do this, as having both the onchain version and the fixed pallet version is not optimal.

Huh, I never thought about that.  
So how would we ensure that the latest migration is not forgotten to be included in the runtime?  
By exposing some new `Migrations` type in each pallet that is then picked up by the migrations pallet?  
Does sound more complicated than the current approach to me… with the new checks we already have the guarantee that no migration can be forgotten to be included. Although i am not sure how this check behaves in the presence of MBMs.

>>I think it is necessary to keep so that we may correctly set the initial on-chain version of a pallet when it is first initialized.

>can be solved in various other ways.

AFAIK the current was is that there is no version until one is needed. So it only has to be set once we actually need it for a migration.

liamaharon commented:
Got it. Yeah `pallet::storage_version` is only used in `VersionedRuntimeUpgrade` as a trivial sanity check in try-runtime pre-upgrade. I'll just remove the check since we plan on deprecating `pallet::storage_version`, that way there'll be no dependency.

liamaharon committed:
remove pre_upgrade current storage version check

liamaharon committed:
derive_impl

bkchr commented:
> * In the code-base, there is an ever-growing list of migrations. Imagine all migrations from all pallets are always passed to `Executive`, for as long as we maintain them.

This means we add tons of junk code to the runtime of which 99% isn't needed anymore. I don't see any reason why we should do this. This just increases the runtime size for nothing. 

It would be much better to write some tool that generate you the list of required migrations than doing this, IMO.


> * on initialization, use the `To` version on the latest migration to init the on-chain version

This just a bad "magical trick". You pretend to remove something "storage_version", but at the same time introduce the requirement to declare the latest migration. Is see no difference.


> * since migrations are tightly coupled with their pallets, the Migrations pallet could automatically check and execute these upgrades rather than us needing to list them in the `Executive` `Migrations` tuple

I didn't yet looked into what @ggwpez is doing. However, I have the hope that this is just an extension to the minimal migration system. I mean we can go this route, but then this new pallet isn't a pallet and more "frame-migration-system" that is an integral part of a runtime like `frame-system`.

liamaharon commented:
> This means we add tons of junk code to the runtime of which 99% isn't needed anymore. I don't see any reason why we should do this. This just increases the runtime size for nothing.

We would only need to retain the most recent migration in the runtime

> This just a bad "magical trick". You pretend to remove something "storage_version", but at the same time introduce the requirement to declare the latest migration. Is see no difference.

I think a difference (and what I think Kian is hoping to achieve) is removing the requirement for us to maintain the latest storage version in two places: once in the pallet, once in the latest migration. Instead, the pallet storage version could be inferred from the latest migration and we wouldn't have the annoying issue of needing to bump it anymore in the pallet, or people forgetting to bump it when they write a migration - single source of truth for the version. 

However there are tradeoffs to consider, including: it could potentially result in misconfiguration where teams unfamiliar with this system don't realise they need to include the latest migration when instantiating a pallet, causing their runtime to incorrectly init the on-chain version to 0 instead of N. In that regard, keeping `pallet::storage_version` + including strong safeguards (like failing `on-runtime-upgrade` if `pallet::storage_version` is < a migration `To` version) may be simpler to understand and harder to mess up, but not as theoretically as clean as it could be.

> I didn't yet looked into what @ggwpez is doing. However, I have the hope that this is just an extension to the minimal migration system. I mean we can go this route, but then this new pallet isn't a pallet and more "frame-migration-system" that is an integral part of a runtime like frame-system.

This was just a random suggestion about how it could theoretically work, it's def not planned.

ggwpez commented:
I dont think `pallet::storage_version` should be removed, and I dont think that is what @kianenigma meant.

liamaharon commented:
> I dont think that is what @kianenigma meant, or?

It was first raised as something to consider here https://github.com/paritytech/substrate/issues/13107#issuecomment-1488548489, and again in this PR https://github.com/paritytech/substrate/pull/14311#issuecomment-1586860496

kianenigma commented:
> This means we add tons of junk code to the runtime of which 99% isn't needed anymore. I don't see any reason why we should do this. This just increases the runtime size for nothing.



My aspiration in keeping this "junk" around is that at least from what I recall, maintaining the "correct list of migrations" for each runtime/release was rather hard. I see that nowadays we have a `type Unreleased = (...)` that is kinda better. Perhaps I am acting too much on the PTSD of the old days :)) 

> I dont think pallet::storage_version should be removed, and I dont think that is what @kianenigma meant, or?


I am indeed convinced that `storage_version` should stay, to be best of my knowledge.

ggwpez commented:
> I see that nowadays we have a type Unreleased = (...) that is kinda better. Perhaps I am acting too much on the PTSD of the old days :))

Yea we can delete them once they were enacted on-chain.  
That is what this `Unreleased` etc helps with. So we can check the on-chain version and then clean up all the old ones.   
The difference is that we remove them after enactment - not after a release.

liamaharon committed:
skip pre/post checks if the actual migration will not run

liamaharon committed:
improve variable naming

liamaharon committed:
docs

ggwpez commented on frame/support/src/migrations.rs:
```diff
@@ -15,19 +15,128 @@
+	/// Executes post_upgrade if the migration will run.
+	#[cfg(feature = "try-runtime")]
+	fn post_upgrade(pre_upgrade_return_bytes: Vec<u8>) -> Result<(), sp_runtime::TryRuntimeError> {
+		let on_chain_version = Pallet::on_chain_storage_version();
+		if on_chain_version == From::get() {
```
Is this correct? Should the on-chain version not be updated afterwards?  
Maybe you can just encode something like `(bool, inner)` and set the bool to whether `pre_upgrade` executed.

liamaharon replied:
Good catch - thanks.

liamaharon committed:
fix post_upgrade 'should run' logic

liamaharon committed:
fix comments

ggwpez commented on frame/support/src/migrations.rs:
```diff
@@ -129,9 +131,13 @@ impl<
-			Inner::post_upgrade(pre_upgrade_return_bytes)
+		use codec::Decode;
+		let (inner_bytes, migration_ran) =
+			<(Vec<u8>, bool)>::decode(&mut &pre_upgrade_return_bytes[..]).map_err(|_| {
```
```suggestion
			<(Vec<u8>, bool)>::decode_all(&mut &pre_upgrade_return_bytes[..]).map_err(|_| {
```
Just to be sure.

ggwpez commented on frame/support/src/migrations.rs:
```diff
@@ -15,19 +15,135 @@
+		if on_chain_version == From::get() {
+			Ok((Inner::pre_upgrade(), true).encode())
+		} else {
+			Ok((Vec::<u8>::new(), false).encode())
```
Some log that it was skipped?

liamaharon replied:
A log will always fire in the `on_runtime_upgrade`. I don't think it's necessary to duplicate the log, but lmk if there's a reason I'm missing.

ggwpez commented on frame/support/src/migrations.rs:
```diff
@@ -15,19 +15,135 @@
+		use codec::Encode;
+		let on_chain_version = Pallet::on_chain_storage_version();
+		if on_chain_version == From::get() {
+			Ok((Inner::pre_upgrade(), true).encode())
```
`pre_upgrade` must not alter storage. AFAIK we already check that in FRAME, can you double-check please if that is the case?  
Otherwise we would need a `StorageNoopGuard` here. Same for `post_upgrade`.

liamaharon replied:
haha I was literally just writing a test trying to lazy and store if pre/post hooks had been called by altering storage and bumped into that `StorageNoopGuard` error, was wondering what that was.. I'll use a different method to test that pre/post hooks are called/not called as expected :)

ggwpez commented on frame/support/tests/versioned_runtime_upgrade.rs:
```diff
@@ -0,0 +1,208 @@
+// Because `derive_impl` is a [macro_magic](https://crates.io/crates/macro_magic) attribute
+// macro, [`#[use_attr]`](`frame_support::macro_magic::use_attr`) must be attached to any use
+// statement that brings it into scope.
+#[use_attr]
```
I think this got removed, or @sam0x17 ?

ggwpez commented on frame/support/tests/versioned_runtime_upgrade.rs:
```diff
@@ -0,0 +1,208 @@
+	}
+}
+
+type VersionedMigrationV0ToV1 = VersionedRuntimeUpgrade<
+	ConstU16<0>,
+	ConstU16<1>,
+	SomeUnversionedMigration<Test, ConstU32<1>>,
+	DummyPallet,
+	RocksDbWeight,
+>;
```
I think this looks good. Associated consts would actually make it more complicated, or?  
But instead of `ConstU16` it should hopefully be possible to use `const u16` as Rust generic.

liamaharon replied:
`const u16` seems to work. The only potential issue is Rust wants me to uppercase them, so we'd have FROM and TO instead of From and To. WDYT?

ggwpez replied:
I think we only used one letter generics for types; `FROM` and `TO` is good.

liamaharon replied:
Refactored to use native rust generics for FROM and TO

ggwpez reviewed (approved):
Looks mostly good. Probably have to try it to find the sharp edges.  
Do you think we can make this `experimental`?

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3035674

liamaharon commented:
> Looks mostly good. Probably have to try it to find the sharp edges.
> Do you think we can make this experimental?

Sure, have we decided on an experimental feature flag?

liamaharon committed:
pre and post hook tests

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3036004

liamaharon committed:
feature gate try-runtime stuff

liamaharon committed:
Merge branch 'master' of github.com:paritytech/substrate into liam-versioned-runtime-upgrade

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3036169

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3036170

liamaharon committed:
remove deprecated macro

ggwpez commented:
> Sure, have we decided on an experimental feature flag?

https://github.com/paritytech/substrate/issues/14345#issuecomment-1594827630

liamaharon committed:
Update frame/support/src/migrations.rs

Co-authored-by: Oliver Tale-Yazdi <oliver.tale-yazdi@parity.io>

liamaharon committed:
decode_all

liamaharon committed:
Merge branch 'liam-versioned-runtime-upgrade' of github.com:paritytech/substrate into liam-versioned-runtime-upgrade

liamaharon committed:
make experimental

liamaharon commented:
@ggwpez could you please double check I set this as an experimental feature correctly?

liamaharon commented on frame/support/tests/versioned_runtime_upgrade.rs:
```diff
@@ -0,0 +1,259 @@
+	const UpgradeWrites: u64 = 2;
+}
+
+#[cfg(feature = "try-runtime")]
+static PRE_UPGRADE_RETURN_BYTES: [u8; 4] = [0, 1, 2, 3];
+
+// We can't write to pallet storage from pre/post hooks, so use a global variable to track that they
+// are called correctly.
+#[cfg(feature = "try-runtime")]
+static PRE_UPGRADE_CALLED: Lazy<Mutex<bool>> = Lazy::new(|| Mutex::new(false));
+#[cfg(feature = "try-runtime")]
+static POST_UPGRADE_CALLED: Lazy<Mutex<bool>> = Lazy::new(|| Mutex::new(false));
+#[cfg(feature = "try-runtime")]
+static POST_UPGRADE_CALLED_WITH: Lazy<Mutex<Vec<u8>>> = Lazy::new(|| Mutex::new(Vec::new()));
```
This is pretty messy. Any suggestions for not needing to repeat `#[cfg(feature = "try-runtime")]`?

ggwpez replied:
You can put it into a `mod`, or just do `allow_unsed` or whatever it reports.

liamaharon committed:
use rust generics

ggwpez commented on frame/support/tests/versioned_runtime_upgrade.rs:
```diff
@@ -0,0 +1,243 @@
+// We can't write to pallet storage from pre/post hooks, so use a global variable to track that they
+// are called correctly.
+#[cfg(feature = "try-runtime")]
+static PRE_UPGRADE_CALLED: Lazy<Mutex<bool>> = Lazy::new(|| Mutex::new(false));
```
This should not be multi threaded?!  
`parameter_types!` can also generate thread locals, [example](https://github.com/paritytech/substrate/blob/8fca6cf78642ebaaa3dfcb306574d2d503a36bd9/frame/message-queue/src/mock.rs#L97).

liamaharon replied:
Ah cool, I did not know that we could `::set` a parameter type. Will update.

ggwpez commented on frame/support/tests/versioned_runtime_upgrade.rs:
```diff
@@ -0,0 +1,243 @@
+
+//! Tests for VersionedRuntimeUpgrade
+
+#![cfg(feature = "experimental")]
```
Could also opt to add `try-runtime` here, then there is no problem with the imports.

liamaharon replied:
Do you know if `try-runtime` tests are run in the CI? If so, then I see no issue with that

ggwpez replied:
Yea they should.

ggwpez reviewed (approved):
The feature looks fine.

liamaharon committed:
add info log when running

liamaharon committed:
simplify tests

liamaharon committed:
improve log

liamaharon committed:
improve log

liamaharon committed:
cleaner pre_upgrade encoding

ggwpez commented:
I think we still use the `C*` labels for release priority. So anything that is not a hot/security fix is "low".

gpestana commented on frame/support/src/migrations.rs:
```diff
@@ -25,9 +23,138 @@ use impl_trait_for_tuples::impl_for_tuples;
+		if on_chain_version == FROM {
+			Ok(Some(Inner::pre_upgrade()?).encode())
+		} else {
+			Ok(None::<Vec<u8>>.encode())
```
As we discussed, maybe returning an encoded enum variant that may wrap the return vec u8 of the `Inner::pre_upgrade` would be a good approach, especially from an ergonomics perspective.

```rust
pub enum PreUpgradeData {
  UpgradeVersionsMismatch,
  Data(Vec<u8>)
} 
```

liamaharon committed:
Update frame/support/src/migrations.rs

Co-authored-by: Gonçalo Pestana <g6pestana@gmail.com>

liamaharon committed:
Update frame/support/src/migrations.rs

Co-authored-by: Gonçalo Pestana <g6pestana@gmail.com>

liamaharon committed:
Update frame/support/src/migrations.rs

Co-authored-by: Gonçalo Pestana <g6pestana@gmail.com>

liamaharon committed:
Update frame/support/src/migrations.rs

Co-authored-by: Gonçalo Pestana <g6pestana@gmail.com>

liamaharon committed:
Update frame/support/src/migrations.rs

Co-authored-by: Gonçalo Pestana <g6pestana@gmail.com>

liamaharon committed:
Merge branch 'master' of github.com:paritytech/substrate into liam-versioned-runtime-upgrade

liamaharon committed:
VersionedPostUpgradeData enum

liamaharon committed:
move versioned runtime upgrade tests to test/tests

liamaharon committed:
fix rust doc

liamaharon committed:
Merge branch 'master' of github.com:paritytech/substrate into liam-versioned-runtime-upgrade

liamaharon committed:
clarify comment

liamaharon commented:
bot merge
